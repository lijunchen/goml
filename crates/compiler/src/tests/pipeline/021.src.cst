FILE@0..846
  FN@0..88
    FnKeyword@0..2 "fn"
    Whitespace@2..3 " "
    Lident@3..12 "match_int"
    PARAM_LIST@12..21
      LParen@12..13 "("
      PARAM@13..19
        Lident@13..14 "n"
        Colon@14..15 ":"
        Whitespace@15..16 " "
        TYPE_INT@16..19
          IntKeyword@16..19 "int"
      RParen@19..20 ")"
      Whitespace@20..21 " "
    Arrow@21..23 "->"
    Whitespace@23..24 " "
    TYPE_INT@24..28
      IntKeyword@24..27 "int"
      Whitespace@27..28 " "
    BLOCK@28..88
      LBrace@28..29 "{"
      Whitespace@29..32 "\n  "
      EXPR_MATCH@32..85
        MatchKeyword@32..37 "match"
        Whitespace@37..38 " "
        EXPR_LIDENT@38..40
          Lident@38..39 "n"
          Whitespace@39..40 " "
        MATCH_ARM_LIST@40..85
          LBrace@40..41 "{"
          Whitespace@41..46 "\n    "
          MATCH_ARM@46..53
            PATTERN_INT@46..48
              Int@46..47 "0"
              Whitespace@47..48 " "
            FatArrow@48..50 "=>"
            Whitespace@50..51 " "
            EXPR_INT@51..53
              Int@51..53 "10"
          Comma@53..54 ","
          Whitespace@54..59 "\n    "
          MATCH_ARM@59..66
            PATTERN_INT@59..61
              Int@59..60 "1"
              Whitespace@60..61 " "
            FatArrow@61..63 "=>"
            Whitespace@63..64 " "
            EXPR_INT@64..66
              Int@64..66 "20"
          Comma@66..67 ","
          Whitespace@67..72 "\n    "
          MATCH_ARM@72..79
            PATTERN_WILDCARD@72..74
              WildcardKeyword@72..73 "_"
              Whitespace@73..74 " "
            FatArrow@74..76 "=>"
            Whitespace@76..77 " "
            EXPR_INT@77..79
              Int@77..79 "30"
          Comma@79..80 ","
          Whitespace@80..83 "\n  "
          RBrace@83..84 "}"
          Whitespace@84..85 "\n"
      RBrace@85..86 "}"
      Whitespace@86..88 "\n\n"
  FN@88..168
    FnKeyword@88..90 "fn"
    Whitespace@90..91 " "
    Lident@91..105 "wildcard_first"
    PARAM_LIST@105..114
      LParen@105..106 "("
      PARAM@106..112
        Lident@106..107 "n"
        Colon@107..108 ":"
        Whitespace@108..109 " "
        TYPE_INT@109..112
          IntKeyword@109..112 "int"
      RParen@112..113 ")"
      Whitespace@113..114 " "
    Arrow@114..116 "->"
    Whitespace@116..117 " "
    TYPE_INT@117..121
      IntKeyword@117..120 "int"
      Whitespace@120..121 " "
    BLOCK@121..168
      LBrace@121..122 "{"
      Whitespace@122..125 "\n  "
      EXPR_MATCH@125..165
        MatchKeyword@125..130 "match"
        Whitespace@130..131 " "
        EXPR_LIDENT@131..133
          Lident@131..132 "n"
          Whitespace@132..133 " "
        MATCH_ARM_LIST@133..165
          LBrace@133..134 "{"
          Whitespace@134..139 "\n    "
          MATCH_ARM@139..146
            PATTERN_WILDCARD@139..141
              WildcardKeyword@139..140 "_"
              Whitespace@140..141 " "
            FatArrow@141..143 "=>"
            Whitespace@143..144 " "
            EXPR_INT@144..146
              Int@144..146 "40"
          Comma@146..147 ","
          Whitespace@147..152 "\n    "
          MATCH_ARM@152..159
            PATTERN_INT@152..154
              Int@152..153 "0"
              Whitespace@153..154 " "
            FatArrow@154..156 "=>"
            Whitespace@156..157 " "
            EXPR_INT@157..159
              Int@157..159 "50"
          Comma@159..160 ","
          Whitespace@160..163 "\n  "
          RBrace@163..164 "}"
          Whitespace@164..165 "\n"
      RBrace@165..166 "}"
      Whitespace@166..168 "\n\n"
  FN@168..264
    FnKeyword@168..170 "fn"
    Whitespace@170..171 " "
    Lident@171..186 "wildcard_middle"
    PARAM_LIST@186..195
      LParen@186..187 "("
      PARAM@187..193
        Lident@187..188 "n"
        Colon@188..189 ":"
        Whitespace@189..190 " "
        TYPE_INT@190..193
          IntKeyword@190..193 "int"
      RParen@193..194 ")"
      Whitespace@194..195 " "
    Arrow@195..197 "->"
    Whitespace@197..198 " "
    TYPE_INT@198..202
      IntKeyword@198..201 "int"
      Whitespace@201..202 " "
    BLOCK@202..264
      LBrace@202..203 "{"
      Whitespace@203..206 "\n  "
      EXPR_MATCH@206..261
        MatchKeyword@206..211 "match"
        Whitespace@211..212 " "
        EXPR_LIDENT@212..214
          Lident@212..213 "n"
          Whitespace@213..214 " "
        MATCH_ARM_LIST@214..261
          LBrace@214..215 "{"
          Whitespace@215..220 "\n    "
          MATCH_ARM@220..227
            PATTERN_INT@220..222
              Int@220..221 "2"
              Whitespace@221..222 " "
            FatArrow@222..224 "=>"
            Whitespace@224..225 " "
            EXPR_INT@225..227
              Int@225..227 "90"
          Comma@227..228 ","
          Whitespace@228..233 "\n    "
          MATCH_ARM@233..241
            PATTERN_WILDCARD@233..235
              WildcardKeyword@233..234 "_"
              Whitespace@234..235 " "
            FatArrow@235..237 "=>"
            Whitespace@237..238 " "
            EXPR_INT@238..241
              Int@238..241 "100"
          Comma@241..242 ","
          Whitespace@242..247 "\n    "
          MATCH_ARM@247..255
            PATTERN_INT@247..249
              Int@247..248 "3"
              Whitespace@248..249 " "
            FatArrow@249..251 "=>"
            Whitespace@251..252 " "
            EXPR_INT@252..255
              Int@252..255 "110"
          Comma@255..256 ","
          Whitespace@256..259 "\n  "
          RBrace@259..260 "}"
          Whitespace@260..261 "\n"
      RBrace@261..262 "}"
      Whitespace@262..264 "\n\n"
  FN@264..351
    FnKeyword@264..266 "fn"
    Whitespace@266..267 " "
    Lident@267..275 "repeated"
    PARAM_LIST@275..284
      LParen@275..276 "("
      PARAM@276..282
        Lident@276..277 "n"
        Colon@277..278 ":"
        Whitespace@278..279 " "
        TYPE_INT@279..282
          IntKeyword@279..282 "int"
      RParen@282..283 ")"
      Whitespace@283..284 " "
    Arrow@284..286 "->"
    Whitespace@286..287 " "
    TYPE_INT@287..291
      IntKeyword@287..290 "int"
      Whitespace@290..291 " "
    BLOCK@291..351
      LBrace@291..292 "{"
      Whitespace@292..295 "\n  "
      EXPR_MATCH@295..348
        MatchKeyword@295..300 "match"
        Whitespace@300..301 " "
        EXPR_LIDENT@301..303
          Lident@301..302 "n"
          Whitespace@302..303 " "
        MATCH_ARM_LIST@303..348
          LBrace@303..304 "{"
          Whitespace@304..309 "\n    "
          MATCH_ARM@309..316
            PATTERN_INT@309..311
              Int@309..310 "1"
              Whitespace@310..311 " "
            FatArrow@311..313 "=>"
            Whitespace@313..314 " "
            EXPR_INT@314..316
              Int@314..316 "60"
          Comma@316..317 ","
          Whitespace@317..322 "\n    "
          MATCH_ARM@322..329
            PATTERN_INT@322..324
              Int@322..323 "1"
              Whitespace@323..324 " "
            FatArrow@324..326 "=>"
            Whitespace@326..327 " "
            EXPR_INT@327..329
              Int@327..329 "70"
          Comma@329..330 ","
          Whitespace@330..335 "\n    "
          MATCH_ARM@335..342
            PATTERN_WILDCARD@335..337
              WildcardKeyword@335..336 "_"
              Whitespace@336..337 " "
            FatArrow@337..339 "=>"
            Whitespace@339..340 " "
            EXPR_INT@340..342
              Int@340..342 "80"
          Comma@342..343 ","
          Whitespace@343..346 "\n  "
          RBrace@346..347 "}"
          Whitespace@347..348 "\n"
      RBrace@348..349 "}"
      Whitespace@349..351 "\n\n"
  FN@351..846
    FnKeyword@351..353 "fn"
    Whitespace@353..354 " "
    Lident@354..358 "main"
    PARAM_LIST@358..361
      LParen@358..359 "("
      RParen@359..360 ")"
      Whitespace@360..361 " "
    BLOCK@361..846
      LBrace@361..362 "{"
      Whitespace@362..367 "\n    "
      EXPR_LET@367..844
        LetKeyword@367..370 "let"
        Whitespace@370..371 " "
        PATTERN_WILDCARD@371..373
          WildcardKeyword@371..372 "_"
          Whitespace@372..373 " "
        Eq@373..374 "="
        Whitespace@374..375 " "
        EXPR_LET_VALUE@375..419
          EXPR_CALL@375..419
            EXPR_LIDENT@375..389
              Lident@375..389 "string_println"
            ARG_LIST@389..419
              LParen@389..390 "("
              ARG@390..417
                EXPR_CALL@390..417
                  EXPR_LIDENT@390..403
                    Lident@390..403 "int_to_string"
                  ARG_LIST@403..417
                    LParen@403..404 "("
                    ARG@404..416
                      EXPR_CALL@404..416
                        EXPR_LIDENT@404..413
                          Lident@404..413 "match_int"
                        ARG_LIST@413..416
                          LParen@413..414 "("
                          ARG@414..415
                            EXPR_INT@414..415
                              Int@414..415 "0"
                          RParen@415..416 ")"
                    RParen@416..417 ")"
              RParen@417..418 ")"
              Whitespace@418..419 " "
        InKeyword@419..421 "in"
        Whitespace@421..426 "\n    "
        EXPR_LET_BODY@426..844
          EXPR_LET@426..844
            LetKeyword@426..429 "let"
            Whitespace@429..430 " "
            PATTERN_WILDCARD@430..432
              WildcardKeyword@430..431 "_"
              Whitespace@431..432 " "
            Eq@432..433 "="
            Whitespace@433..434 " "
            EXPR_LET_VALUE@434..478
              EXPR_CALL@434..478
                EXPR_LIDENT@434..448
                  Lident@434..448 "string_println"
                ARG_LIST@448..478
                  LParen@448..449 "("
                  ARG@449..476
                    EXPR_CALL@449..476
                      EXPR_LIDENT@449..462
                        Lident@449..462 "int_to_string"
                      ARG_LIST@462..476
                        LParen@462..463 "("
                        ARG@463..475
                          EXPR_CALL@463..475
                            EXPR_LIDENT@463..472
                              Lident@463..472 "match_int"
                            ARG_LIST@472..475
                              LParen@472..473 "("
                              ARG@473..474
                                EXPR_INT@473..474
                                  Int@473..474 "5"
                              RParen@474..475 ")"
                        RParen@475..476 ")"
                  RParen@476..477 ")"
                  Whitespace@477..478 " "
            InKeyword@478..480 "in"
            Whitespace@480..485 "\n    "
            EXPR_LET_BODY@485..844
              EXPR_LET@485..844
                LetKeyword@485..488 "let"
                Whitespace@488..489 " "
                PATTERN_WILDCARD@489..491
                  WildcardKeyword@489..490 "_"
                  Whitespace@490..491 " "
                Eq@491..492 "="
                Whitespace@492..493 " "
                EXPR_LET_VALUE@493..542
                  EXPR_CALL@493..542
                    EXPR_LIDENT@493..507
                      Lident@493..507 "string_println"
                    ARG_LIST@507..542
                      LParen@507..508 "("
                      ARG@508..540
                        EXPR_CALL@508..540
                          EXPR_LIDENT@508..521
                            Lident@508..521 "int_to_string"
                          ARG_LIST@521..540
                            LParen@521..522 "("
                            ARG@522..539
                              EXPR_CALL@522..539
                                EXPR_LIDENT@522..536
                                  Lident@522..536 "wildcard_first"
                                ARG_LIST@536..539
                                  LParen@536..537 "("
                                  ARG@537..538
                                    EXPR_INT@537..538
                                      Int@537..538 "0"
                                  RParen@538..539 ")"
                            RParen@539..540 ")"
                      RParen@540..541 ")"
                      Whitespace@541..542 " "
                InKeyword@542..544 "in"
                Whitespace@544..549 "\n    "
                EXPR_LET_BODY@549..844
                  EXPR_LET@549..844
                    LetKeyword@549..552 "let"
                    Whitespace@552..553 " "
                    PATTERN_WILDCARD@553..555
                      WildcardKeyword@553..554 "_"
                      Whitespace@554..555 " "
                    Eq@555..556 "="
                    Whitespace@556..557 " "
                    EXPR_LET_VALUE@557..606
                      EXPR_CALL@557..606
                        EXPR_LIDENT@557..571
                          Lident@557..571 "string_println"
                        ARG_LIST@571..606
                          LParen@571..572 "("
                          ARG@572..604
                            EXPR_CALL@572..604
                              EXPR_LIDENT@572..585
                                Lident@572..585 "int_to_string"
                              ARG_LIST@585..604
                                LParen@585..586 "("
                                ARG@586..603
                                  EXPR_CALL@586..603
                                    EXPR_LIDENT@586..600
                                      Lident@586..600 "wildcard_first"
                                    ARG_LIST@600..603
                                      LParen@600..601 "("
                                      ARG@601..602
                                        EXPR_INT@601..602
                                          Int@601..602 "2"
                                      RParen@602..603 ")"
                                RParen@603..604 ")"
                          RParen@604..605 ")"
                          Whitespace@605..606 " "
                    InKeyword@606..608 "in"
                    Whitespace@608..613 "\n    "
                    EXPR_LET_BODY@613..844
                      EXPR_LET@613..844
                        LetKeyword@613..616 "let"
                        Whitespace@616..617 " "
                        PATTERN_WILDCARD@617..619
                          WildcardKeyword@617..618 "_"
                          Whitespace@618..619 " "
                        Eq@619..620 "="
                        Whitespace@620..621 " "
                        EXPR_LET_VALUE@621..671
                          EXPR_CALL@621..671
                            EXPR_LIDENT@621..635
                              Lident@621..635 "string_println"
                            ARG_LIST@635..671
                              LParen@635..636 "("
                              ARG@636..669
                                EXPR_CALL@636..669
                                  EXPR_LIDENT@636..649
                                    Lident@636..649 "int_to_string"
                                  ARG_LIST@649..669
                                    LParen@649..650 "("
                                    ARG@650..668
                                      EXPR_CALL@650..668
                                        EXPR_LIDENT@650..665
                                          Lident@650..665 "wildcard_middle"
                                        ARG_LIST@665..668
                                          LParen@665..666 "("
                                          ARG@666..667
                                            EXPR_INT@666..667
                                              Int@666..667 "2"
                                          RParen@667..668 ")"
                                    RParen@668..669 ")"
                              RParen@669..670 ")"
                              Whitespace@670..671 " "
                        InKeyword@671..673 "in"
                        Whitespace@673..678 "\n    "
                        EXPR_LET_BODY@678..844
                          EXPR_LET@678..844
                            LetKeyword@678..681 "let"
                            Whitespace@681..682 " "
                            PATTERN_WILDCARD@682..684
                              WildcardKeyword@682..683 "_"
                              Whitespace@683..684 " "
                            Eq@684..685 "="
                            Whitespace@685..686 " "
                            EXPR_LET_VALUE@686..736
                              EXPR_CALL@686..736
                                EXPR_LIDENT@686..700
                                  Lident@686..700 "string_println"
                                ARG_LIST@700..736
                                  LParen@700..701 "("
                                  ARG@701..734
                                    EXPR_CALL@701..734
                                      EXPR_LIDENT@701..714
                                        Lident@701..714 "int_to_string"
                                      ARG_LIST@714..734
                                        LParen@714..715 "("
                                        ARG@715..733
                                          EXPR_CALL@715..733
                                            EXPR_LIDENT@715..730
                                              Lident@715..730 "wildcard_middle"
                                            ARG_LIST@730..733
                                              LParen@730..731 "("
                                              ARG@731..732
                                                EXPR_INT@731..732
                                                  Int@731..732 "3"
                                              RParen@732..733 ")"
                                        RParen@733..734 ")"
                                  RParen@734..735 ")"
                                  Whitespace@735..736 " "
                            InKeyword@736..738 "in"
                            Whitespace@738..743 "\n    "
                            EXPR_LET_BODY@743..844
                              EXPR_LET@743..844
                                LetKeyword@743..746 "let"
                                Whitespace@746..747 " "
                                PATTERN_WILDCARD@747..749
                                  WildcardKeyword@747..748 "_"
                                  Whitespace@748..749 " "
                                Eq@749..750 "="
                                Whitespace@750..751 " "
                                EXPR_LET_VALUE@751..794
                                  EXPR_CALL@751..794
                                    EXPR_LIDENT@751..765
                                      Lident@751..765 "string_println"
                                    ARG_LIST@765..794
                                      LParen@765..766 "("
                                      ARG@766..792
                                        EXPR_CALL@766..792
                                          EXPR_LIDENT@766..779
                                            Lident@766..779 "int_to_string"
                                          ARG_LIST@779..792
                                            LParen@779..780 "("
                                            ARG@780..791
                                              EXPR_CALL@780..791
                                                EXPR_LIDENT@780..788
                                                  Lident@780..788 "repeated"
                                                ARG_LIST@788..791
                                                  LParen@788..789 "("
                                                  ARG@789..790
                                                    EXPR_INT@789..790
                                                      Int@789..790 "1"
                                                  RParen@790..791 ")"
                                            RParen@791..792 ")"
                                      RParen@792..793 ")"
                                      Whitespace@793..794 " "
                                InKeyword@794..796 "in"
                                Whitespace@796..801 "\n    "
                                EXPR_LET_BODY@801..844
                                  EXPR_CALL@801..844
                                    EXPR_LIDENT@801..815
                                      Lident@801..815 "string_println"
                                    ARG_LIST@815..844
                                      LParen@815..816 "("
                                      ARG@816..842
                                        EXPR_CALL@816..842
                                          EXPR_LIDENT@816..829
                                            Lident@816..829 "int_to_string"
                                          ARG_LIST@829..842
                                            LParen@829..830 "("
                                            ARG@830..841
                                              EXPR_CALL@830..841
                                                EXPR_LIDENT@830..838
                                                  Lident@830..838 "repeated"
                                                ARG_LIST@838..841
                                                  LParen@838..839 "("
                                                  ARG@839..840
                                                    EXPR_INT@839..840
                                                      Int@839..840 "3"
                                                  RParen@840..841 ")"
                                            RParen@841..842 ")"
                                      RParen@842..843 ")"
                                      Whitespace@843..844 "\n"
      RBrace@844..845 "}"
      Whitespace@845..846 "\n"