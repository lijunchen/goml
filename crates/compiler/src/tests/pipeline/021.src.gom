package main

import (
    "fmt"
)

func int_to_string(x int) string {
    return fmt.Sprintf("%d", x)
}

func string_println(s string) struct{} {
    fmt.Println(s)
    return struct{}{}
}

func match_int(n__0 int) int {
    var ret23 int
    switch n__0 {
    case 0:
        ret23 = 10
    case 1:
        ret23 = 20
    default:
        ret23 = 30
    }
    return ret23
}

func wildcard_first(n__1 int) int {
    var ret24 int
    ret24 = 40
    return ret24
}

func wildcard_middle(n__2 int) int {
    var ret25 int
    switch n__2 {
    case 2:
        ret25 = 90
    case 3:
        ret25 = 100
    default:
        ret25 = 100
    }
    return ret25
}

func repeated(n__3 int) int {
    var ret26 int
    switch n__3 {
    case 1:
        ret26 = 60
    default:
        ret26 = 80
    }
    return ret26
}

func main0() struct{} {
    var ret27 struct{}
    var t8 int = match_int(0)
    var t7 string = int_to_string(t8)
    string_println(t7)
    var t10 int = match_int(5)
    var t9 string = int_to_string(t10)
    string_println(t9)
    var t12 int = wildcard_first(0)
    var t11 string = int_to_string(t12)
    string_println(t11)
    var t14 int = wildcard_first(2)
    var t13 string = int_to_string(t14)
    string_println(t13)
    var t16 int = wildcard_middle(2)
    var t15 string = int_to_string(t16)
    string_println(t15)
    var t18 int = wildcard_middle(3)
    var t17 string = int_to_string(t18)
    string_println(t17)
    var t20 int = repeated(1)
    var t19 string = int_to_string(t20)
    string_println(t19)
    var t22 int = repeated(3)
    var t21 string = int_to_string(t22)
    ret27 = string_println(t21)
    return ret27
}

func main() {
    main0()
}
