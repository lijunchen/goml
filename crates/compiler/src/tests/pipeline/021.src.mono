fn match_int(n/0: int) -> int {
  match n/0 {
    0 => {
      10
    },
    1 => {
      20
    },
    _ => 30,
  }
}

fn wildcard_first(n/1: int) -> int {
  40
}

fn wildcard_middle(n/2: int) -> int {
  match n/2 {
    2 => {
      90
    },
    3 => {
      100
    },
    _ => 100,
  }
}

fn repeated(n/3: int) -> int {
  match n/3 {
    1 => {
      60
    },
    _ => 80,
  }
}

fn main() -> unit {
  let mtmp0 = string_println(int_to_string(match_int(0))) in
  let mtmp1 = string_println(int_to_string(match_int(5))) in
  let mtmp2 = string_println(int_to_string(wildcard_first(0))) in
  let mtmp3 = string_println(int_to_string(wildcard_first(2))) in
  let mtmp4 = string_println(int_to_string(wildcard_middle(2))) in
  let mtmp5 = string_println(int_to_string(wildcard_middle(3))) in
  let mtmp6 = string_println(int_to_string(repeated(1))) in
  string_println(int_to_string(repeated(3)))
}