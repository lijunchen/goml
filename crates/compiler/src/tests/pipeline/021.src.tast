fn match_int(n/0: int) -> int {
  match (n/0 : int) {
      0 => 10,
      1 => 20,
      _ : int => 30,
  }
}

fn wildcard_first(n/1: int) -> int {
  match (n/1 : int) {
      _ : int => 40,
      0 => 50,
  }
}

fn wildcard_middle(n/2: int) -> int {
  match (n/2 : int) {
      2 => 90,
      _ : int => 100,
      3 => 110,
  }
}

fn repeated(n/3: int) -> int {
  match (n/3 : int) {
      1 => 60,
      1 => 70,
      _ : int => 80,
  }
}

fn main() -> unit {
  let _ : unit = string_println(int_to_string(match_int(0))) in
  let _ : unit = string_println(int_to_string(match_int(5))) in
  let _ : unit = string_println(int_to_string(wildcard_first(0))) in
  let _ : unit = string_println(int_to_string(wildcard_first(2))) in
  let _ : unit = string_println(int_to_string(wildcard_middle(2))) in
  let _ : unit = string_println(int_to_string(wildcard_middle(3))) in
  let _ : unit = string_println(int_to_string(repeated(1))) in
  string_println(int_to_string(repeated(3)))
}