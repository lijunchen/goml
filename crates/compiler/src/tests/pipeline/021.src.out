10
30
40
40
90
100
60
80
